lib/galaxy/model/__init__.py

```diff
@@ -4441,6 +4457,227 @@ def __init__(self, user=None, session=None, openid=None):
+
+    def save(self):
+        self.trans.sa_session.add(self)
+        self.trans.sa_session.flush()
+
+    @classmethod
+    def load(cls, token):
+        return cls.trans.sa_session.query(cls).filter(cls.token == token).first()
+
+    @classmethod
+    def destroy(cls, token):
+        partial = cls.load(token)
+        if partial:
+            cls.trans.sa_session.delete(partial)
+
+
+class UserAuthnzToken(UserMixin):
+    __table_args__ = (UniqueConstraint('provider', 'uid'),)
+
+    # This static property is of type: galaxy.web.framework.webapp.GalaxyWebTransaction
+    # and it is set in: galaxy.authnz.psa_authnz.PSAAuthnz
+    trans = None
+
+    def __init__(self, provider, uid, extra_data=None, lifetime=None, assoc_type=None, user=None):
+        self.provider = provider
+        self.uid = uid
+        self.user_id = user.id
+        self.extra_data = extra_data
+        self.lifetime = lifetime
+        self.assoc_type = assoc_type
+
+    def get_id_token(self):
+        return self.extra_data.get('id_token', None) if self.extra_data is not None else None
+
+    def get_access_token(self):
+        return self.extra_data.get('access_token', None) if self.extra_data is not None else None
+
+    def set_extra_data(self, extra_data=None):
+        # Note: the following unicode conversion is a temporary solution for a
+        # database binding error (InterfaceError: (sqlite3.InterfaceError)).
+        if extra_data is not None:
+            extra_data = str(extra_data)
+        if super(UserAuthnzToken, self).set_extra_data(extra_data):
+            self.trans.sa_session.add(self)
+            self.trans.sa_session.flush()
+
+    def save(self):
+        self.trans.sa_session.add(self)
+        self.trans.sa_session.flush()
+
+    @classmethod
+    def username_max_length(cls):
+        # Note: This is the maximum field length set for the username column of the galaxy_user table.
+        # A better alternative is to retrieve this number from the table, instead of this const value.
+        return 255
+
+    @classmethod
+    def user_model(cls):
+        return User
+
+    @classmethod
+    def changed(cls, user):
+        cls.trans.sa_session.add(user)
+        cls.trans.sa_session.flush()
+
+    @classmethod
+    def user_query(cls):
+        return cls.trans.sa_session.query(cls.user_model())
+
+    @classmethod
+    def user_exists(cls, *args, **kwargs):
+        return cls.user_query().filter_by(*args, **kwargs).count() > 0
+
+    @classmethod
+    def get_username(cls, user):
+        return getattr(user, 'username', None)
+
+    @classmethod
+    def create_user(cls, *args, **kwargs):
+        # TODO: password is a required field for a galaxy user record. However, it should not be required
+        # if the user is authenticated by an external identity provider. Once this logic is incorporated
+        # in the galaxy user, the following argument setting should be removed.
+        kwargs['password'] = ''.join(
```

Comment:
Do I understand it correctly that this generated password *cannot* be used to log in?

Reply:
Can be used, but impossible to guess it as it is a random password.

The goal of this PR is to let a user login to Galaxy without having to (explicitly) create a Galaxy user. However, a Galaxy `user` is required to "login" a user. Therefore, a Galaxy `user` must be created implicitly---transparent to the end-user. On the other hand, `password` is a required field to create a Galaxy `user`. If I ask a user to provide a password when he/she is trying to login using third-party identity, then that would not be much different from asking the user to create a Galaxy account. Therefore, a simple approach forward is to create a "random" password, which the user does not need to know because he/she will be able to login with the third-party identity (Google in this case) without entering/creating a username and/or password for Galaxy.

I agree a random password is not the best approach, but tweaking Galaxy's logic on enforcing password as a required field would have made this PR much more complicated. Therefore, I think a random password is a simple approach forward, which can be improved in subsequent and independent PRs.

Reply:
This is still storing a plaintext password in the database though - which is a security problem I think.

Reply:
I can encrypt it like other passwords.

Reply:
To be clear: we store salted hashes, not encrypted passwords. Method used is based upon an admin configurable option. 

However, not generating and allowing a password to be used would be a better approach.

Also might need to consider how this will interact downstream with e.g. upload via ftp.

Reply:
@blankenberg would it be possible for you to elaborate on your concerns for downstream interaction via FTP?

Reply:
@VJalili Our recommended [FTP server configuration](https://docs.galaxyproject.org/en/latest/admin/special_topics/ftp.html) authenticates against Galaxy's DB (so this password would probably be accepted).

Reply:
I can make some tweaks so that: 
a- we still let users login with their third-party identity without explicitly creating a Galaxy user and providing a password
b- for scenarios where Galaxy strictly requires a user password, then a user who is logged-in using third-party identity can still go to a page and set Galaxy password for his/her Galaxy account.